src/util/psbt/map/global.rs

```diff
@@ -158,6 +250,55 @@ impl Decodable for Global {
                                 return Err(Error::InvalidKey(pair.key).into())
                             }
                         }
+                        PSBT_GLOBAL_XPUB => {
+                            if !pair.key.key.is_empty() {
+                                let xpub = ExtendedPubKey::decode(&pair.key.key)
+                                    .map_err(|_| encode::Error::ParseFailed(
+                                        "Can't deserialize ExtendedPublicKey from global XPUB key data"
+                                    ))?;
+
+                                if pair.value.len() % 4 != 0 {
+                                    return Err(encode::Error::ParseFailed("Incorrect length of global xpub list"))
+                                }
+
+                                let keys_count = pair.value.len() / 4 - 1;
+                                let mut decoder = Cursor::new(pair.value);
+                                let mut fingerprint = [0u8; 4];
+                                decoder.read_exact(&mut fingerprint[..])?;
+                                let mut path = Vec::<ChildNumber>::with_capacity(keys_count);
+                                while let Ok(index) = u32::consensus_decode(&mut decoder) {
+                                    path.push(ChildNumber::from(index))
+                                }
+                                let derivation = DerivationPath::from(path);
+                                // Keys, according to BIP-174, must be unique
+                                if xpub_map.insert(xpub, (Fingerprint::from(&fingerprint[..]), derivation)).is_some() {
+                                    return Err(encode::Error::ParseFailed("Repeated global xpub key"))
+                                }
+                            } else {
+                                return Err(encode::Error::ParseFailed("Xpub global key must contain serialized Xpub data"))
+                            }
+                        }
+                        PSBT_GLOBAL_VERSION => {
+                            // key has to be empty
+                            if pair.key.key.is_empty() {
+                                // there can only be one version
+                                if version.is_none() {
+                                    let vlen: usize = pair.value.len();
+                                    let mut decoder = Cursor::new(pair.value);
+                                    if vlen != 4 {
+                                        return Err(encode::Error::ParseFailed("Wrong global version value length (must be 4 bytes)"))
+                                    }
+                                    version = Some(Decodable::consensus_decode(&mut decoder)?);
+                                    if decoder.position() != vlen as u64 {
```

Comment:
Why would a decoder for an `u32` _not_ consume all 4 bytes? The check seems redundant.

Reply:
Fixed in https://github.com/rust-bitcoin/rust-bitcoin/pull/499/commits/b84faa7f5e6f9d273a4a6cb78098f10c73594783